### ISSUE TITLE
Best practices for testing this buildpack locally?

### ISSUE THREAD

**Issue by acechase** [heroku/heroku-buildpack-pgbouncer#44]
I am struggling to find information on how other users of this buildpack are testing the setup locally. 

I have pgbouncer setup on my dev machine (OS X, so I used brew to install and pgbouncer listening on port 6432). I am currently testing by pointing my DB connection at pgbouncer (just changing the port I'm connecting against). But 

I just wanted to check to see if there is some way to better simulate the stunnel wrapper on a local dev machine? I guess trying to simulate a buildpack may be going a step too far?

Figured it was worth creating an issue for this question in case others have it in the future...

**Comment by gregburek** [comment]
Hello, 
I have been able to primarily use a staging app on Heroku to test this buildpack, but I understand that doesn't fit with many work flows. Recently I found a deadlock bug when using this buildpack, sidekiq and sequel where an explicit transaction (`BEGIN` ... `COMMIT`) was blocking the pgbouncer connection and where sequel was unable to spin up a new connection until the explicit connection cleared. 

To find this, I used https://github.com/progrium/buildstep to compile the app locally with debug logging and speculative fixes. I used the generated shell commands to run a sidekiq worker in docker, inside of a boot2docker vm on my laptop. I was able to use `docker exec` for strace, gdb and finally `kill -TTIN $(pgrep ruby)`, which gave me the stack traces that showed the deadlock. My fix was to use `:preconnect` to avoid lazy new connections, as pgbouncer buffers them. 

Does this sound like a helpful workflow for local debugging?

**Comment by acechase** [comment]
@gregburek - Thanks a lot for your reply. 

Buildstep sounds exactly like the kind of thing I need. Our team actually tried to use dokku when we first started deploying test envs to Digital Ocean (we have staging and production on heroku and a number of test envs on Digital Ocean), but we bailed on it at the time because it was problematic getting it to work.

I guess if this continues to be a problem for me I'll start thinking about migrating our other envs to something like Dokku/Buildstep.

Cheers!
-Andrew
